src/daft-core/src/series/array_impl/binary_ops.rs

```diff
@@ -0,0 +1,201 @@
+    datatypes::{Float64Type, Utf8Type},
+    series::series_like::SeriesLike,
+    with_match_numeric_daft_types, DataType,
+};
+
+use crate::datatypes::{
+    BinaryArray, BooleanArray, ExtensionArray, FixedSizeListArray, Float32Array, Float64Array,
+    Int16Array, Int32Array, Int64Array, Int8Array, ListArray, NullArray, StructArray, UInt16Array,
+    UInt32Array, UInt64Array, UInt8Array, Utf8Array,
+};
+
+use crate::datatypes::logical::{
+    DateArray, DurationArray, EmbeddingArray, FixedShapeImageArray, ImageArray, TimestampArray,
+};
+
+use super::{ArrayWrapper, IntoSeries, Series};
+
+#[cfg(feature = "python")]
+use crate::{datatypes::PythonArray, series::ops::py_binary_op_utilfn};
+
+macro_rules! binary_op_default_impl {
+    ($self:expr, $rhs:expr, $op:ident, $default_op:ident) => {{
+        let output_type = ($self.data_type().$op($rhs.data_type()))?;
+        let lhs = $self.into_series();
+        $default_op(&lhs, $rhs, &output_type)
+    }};
+}
+
+pub(crate) trait SeriesBinaryOps: SeriesLike {
+    fn add(&self, rhs: &Series) -> DaftResult<Series> {
+        binary_op_default_impl!(self, rhs, add, physical_add)
+    }
+    fn sub(&self, rhs: &Series) -> DaftResult<Series> {
+        binary_op_default_impl!(self, rhs, sub, physical_sub)
+    }
+    fn mul(&self, rhs: &Series) -> DaftResult<Series> {
+        binary_op_default_impl!(self, rhs, mul, physical_mul)
+    }
+    fn div(&self, rhs: &Series) -> DaftResult<Series> {
+        binary_op_default_impl!(self, rhs, div, physical_div)
+    }
+    fn rem(&self, rhs: &Series) -> DaftResult<Series> {
+        binary_op_default_impl!(self, rhs, rem, physical_rem)
+    }
+}
+
+#[cfg(feature = "python")]
+macro_rules! py_binary_op {
+    ($lhs:expr, $rhs:expr, $pyoperator:expr) => {
+        py_binary_op_utilfn!($lhs, $rhs, $pyoperator, "map_operator_arrow_semantics")
+    };
+}
+
+macro_rules! py_numeric_binary_op {
+    ($op:ident, $pyop:expr, $lhs:expr, $rhs:expr, $output_ty:expr) => {{
+        use DataType::*;
+        match $output_ty {
+            #[cfg(feature = "python")]
+            Python => Ok(py_binary_op!($lhs, $rhs, $pyop)),
+            output_type if output_type.is_numeric() => {
+                let lhs = $lhs.cast(&output_type)?;
+                let rhs = $rhs.cast(&output_type)?;
+                with_match_numeric_daft_types!(output_type, |$T| {
+                    let lhs = lhs.downcast::<$T>()?;
+                    let rhs = rhs.downcast::<$T>()?;
+                    Ok(lhs.$op(rhs)?.into_series().rename(lhs.name()))
+                })
+            }
+            _ => panic!(
+                "No implementation for {} {} {} -> {}",
+                $lhs.data_type(),
+                $pyop,
+                $rhs.data_type(),
+                $output_ty,
+            ),
+        }
+    }};
+}
+
+fn physical_add(lhs: &Series, rhs: &Series, output_type: &DataType) -> DaftResult<Series> {
+    use DataType::*;
+    match output_type {
+        Utf8 => {
```

Comment:
Nice! This is nice because we can match the intended behavior of the Series operation against the intended return type as intended by the type-system, instead of having Series do its own typing logic.

I do wonder about cases where the result dtype isn't fully expressive about the actual runtime implementation though. For example, if the result if a "date", it may not provide enough information about the implementation (is this a date plus a duration or a duration plus a date?)

Reply:
Separately, I guess instead of our current unit tests, we can maybe even just add asserts in our code that will ensure that the series output dtype matches the type-system's intended dtype? Not quite sure how best to tie the resolve-time and runtime behavior together yet.